events/api.py

```
@@ -871,14 +959,27 @@ def create(self, validated_data):
         return event
 
     def update(self, instance, validated_data):
+        # allow updating events if the api key matches event data source
+        if isinstance(self.user, ApiKeyUser):
+            self.user = None
+            if not instance.data_source == self.data_source:
+                raise PermissionDenied()
+        else:
+            if not instance.is_editable() or not instance.is_admin(self.user):
+                raise PermissionDenied()
+
         offers = validated_data.pop('offers', None)
         links = validated_data.pop('external_links', None)
 
+        validated_data['last_modified_by'] = self.user
         # Update event_status if a PUBLIC SCHEDULED or CANCELLED event start_time is updated.
         # DRAFT events will remain SCHEDULED up to publication.
         # Check that the event is not explicitly CANCELLED at the same time.
         if (instance.publication_status == PublicationStatus.PUBLIC and
                     validated_data.get('event_status', Event.Status.SCHEDULED) != Event.Status.CANCELLED):
+            # if the instance was ever RESCHEDULED or POSTPONED, it may never be SCHEDULED again
+            if instance.event_status != Event.Status.SCHEDULED:
+                validated_data['event_status'] = instance.event_status
```

I think it's better to fail with an error instead of silently fixing the data.